File: src/shared/pairing-verification-code.ts
export const PAIRING_VERIFICATION_CODE_LENGTH = 6;

export function createPairingVerificationCode(input: {
  desktopId: string;
  deviceId: string;
  requestNonce: string;
}): string {
  const canonical = `${input.desktopId}\n${input.deviceId}\n${input.requestNonce}`;
  let hash = 2166136261;
  for (const char of canonical) {
    hash ^= char.charCodeAt(0);
    hash = Math.imul(hash, 16777619);
  }
  const value = Math.abs(hash) % 1_000_000;
  return value.toString().padStart(PAIRING_VERIFICATION_CODE_LENGTH, '0');
}

File: src-dotnet/SwitchifyPc.Core/Pairing/PairingVerificationCode.cs
namespace SwitchifyPc.Core.Pairing;

public static class PairingVerificationCode
{
    public const int CodeLength = 6;

    public static string Create(string desktopId, string deviceId, string requestNonce)
    {
        string canonical = $"{desktopId}\n{deviceId}\n{requestNonce}";
        int hash = unchecked((int)2166136261);
        foreach (char character in canonical)
        {
            hash ^= character;
            hash = unchecked(hash * 16777619);
        }

        int value = Math.Abs(hash) % 1_000_000;
        return value.ToString().PadLeft(CodeLength, '0');
    }
}
